wemake_python_styleguide/visitors/ast/statements.py

```diff
@@ -235,3 +254,67 @@ def visit_ClassDef(self, node: ast.ClassDef) -> None:
         all_args = [*node.bases, *[kw.value for kw in node.keywords]]
         self._check_indentation(node, all_args)
         self.generic_visit(node)
+
+
+@final
+@alias('visit_variable_defying_statement', (
+    'visit_With',
+    'visit_AsyncWith',
+    'visit_For',
+    'visit_AsyncFor',
+))
+class VariableUsedOutsideOfBlockVisitor(BaseNodeVisitor):
+    """
+    Responsible for detection of variable defined in `with` and `for` outside of it's body
+    """
+
+    def visit_variable_defying_statement(self, node: AnyVariableDefyingStatement) -> None:
+        """
+        Checks that variable is not used after statement body
+        """
+        node_reached = False
+        defined_names = self._extract_defined_variables(node)
+        if not defined_names:
+            return
+        for statement in node.context.body:
+            if node_reached:
+                self._check_statement(defined_names, statement, node)
+            if node == statement:
+                node_reached = True
+        self.generic_visit(node)
+
+    def _check_statement(self, names: List[str], statement: ast, original_node: AnyVariableDefyingStatement) -> bool:
```

Comment:
It can not be just `statement: ast`. Mypy will fail for this line.
It needs to have a better type.